tests/unit/test_charm_relations.py

```diff
@@ -1,140 +1,149 @@
+
 
+class TestCharmRelations:
+    patcher_check_output = patch("charm.check_output")
+    patcher_generate_csr = patch("charm.generate_csr")
+    patcher_get_assigned_certificates = patch("charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.get_assigned_certificates")  # noqa: E501
+    patcher_request_certificate_creation = patch("charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.request_certificate_creation")  # noqa: E501
 
-class TestCharmRelations(unittest.TestCase):
-    def setUp(self):
-        self.namespace = "whatever"
+    @pytest.fixture()
+    def setup(self):
+        self.mock_check_output = TestCharmRelations.patcher_check_output.start()
+        self.mock_generate_csr = TestCharmRelations.patcher_generate_csr.start()
+        self.mock_get_assigned_certificates = TestCharmRelations.patcher_get_assigned_certificates.start()  # noqa: E501
+        self.mock_request_certificate_creation = TestCharmRelations.patcher_request_certificate_creation.start()  # noqa: E501
+
+    @staticmethod
+    def teardown() -> None:
+        patch.stopall()
+
+    @pytest.fixture(autouse=True)
+    def create_harness(self, setup, request):
         self.harness = testing.Harness(AUSFOperatorCharm)
-        self.harness.set_model_name(name=self.namespace)
-        self.addCleanup(self.harness.cleanup)
+        self.harness.set_model_name(name=NAMESPACE)
         self.harness.set_leader(is_leader=True)
         self.harness.begin()
+        yield self.harness
+        self.harness.cleanup()
+        request.addfinalizer(self.teardown)
 
-    @patch("charm.check_output")
-    @patch(
-        "charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.get_assigned_certificates",  # noqa: E501
-    )
-    def test_given_charm_is_in_active_state_when_certificates_relation_broken_then_certificate_csr_and_private_key_are_removed(  # noqa: E501
-        self, patched_get_assigned_certificates, patched_check_output
-    ):
-        test_private_key = b"whatever private key"
-        test_csr = b"whatever csr"
-        test_certificate = "whatever certificate"
-        self.harness.set_can_connect(container="ausf", val=True)
-        certificates_relation_id, _ = self._create_charm_relations_and_relation_data()
+    @pytest.fixture()
+    def add_storage(self):
         self.harness.add_storage(storage_name="certs", attach=True)
         self.harness.add_storage(storage_name="config", attach=True)
-        patched_check_output.return_value = TEST_POD_IP
-        root = self.harness.get_filesystem_root("ausf")
-        (root / "support/TLS/ausf.csr").write_text(test_csr.decode())
-        (root / "support/TLS/ausf.key").write_text(test_private_key.decode())
-        (root / "support/TLS/ausf.pem").write_text(test_certificate)
+
+    @pytest.fixture()
+    def create_charm_relations_and_relation_data(self) -> Generator[Tuple[int, int], None, None]:
```

Comment:
I am not convinced about the name of the functions and the return values.
In line 46 when the yielded values are actually used, it is not evident that you are getting the relation ID.
I think that I would separate the creation of the certificates relation from the NRF relation into 2 functions (?) and only yield what we actually use